File: src/libres.domain/aggregates/Wallet.ts
import { Aggregate } from "../common/aggregate";
import { Error } from "../common/error";
import { Result } from "../common/result";
import { WalletStatus } from "../enums/wallet-status";

export class Wallet extends Aggregate {

  private _userId: string;
  private _status: WalletStatus = WalletStatus.ACTIVE;
  private _balance: number = 500;
  private _currency: string = "$";
  // private _walletTransactions: WalletTransaction[] = [];
  private _createdAt: Date = new Date();
  private _lastUpdate: Date | null = null;

  protected constructor(id: string, userId: string) {
    super(id);
    this._userId = userId;
  }

  public get userId(): string {
    return this._userId;
  }
  public get status(): WalletStatus {
    return this._status;
  }
  public get balance(): number {
    return this._balance;
  }
  public get currency(): string {
    return this._currency;
  }

  // public get walletTransactions(): ReadonlyArray<WalletTransaction> {
  //     return this._walletTransactions;
  // }
  public get createdAt(): Date {
    return this._createdAt;
  }
  public get lastUpdate(): Date | null {
    return this._lastUpdate;
  }

  public static create(userId: string): Result<Wallet> {
    if (!userId) {
      return Result.Failure<Wallet>(Error.Validation("User id required"));
    }
    const newId = crypto.randomUUID();
    const wallet = new Wallet(newId, userId);
    return Result.Success(wallet);
  }

  public updateLastUpdate(): void {
    this._lastUpdate = new Date();
  }

  public activate(): void {
    this._status = WalletStatus.ACTIVE;
  }

  public freeze(): void {
    this._status = WalletStatus.FREEZE;
  }

  public credit(amount: number): void {
    this._balance += amount;
  }

  public debit(amount: number): Result<string> {
    if (this._balance < amount) {
      return Result.Failure(
        Error.Conflict(`your balance is less than ${amount}`),
      );
    }
    this._balance -= amount;
    return Result.Success("debit Success");
  }
}

File: apps/backend/src/Libres.API/Features/Wallet/Domain/Wallet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Libres.API.Shared.Application.CustomError;
using Libres.API.Shared.Domain;
using Libres.API.Shared.Domain.Enums;

namespace Libres.API.Features.Wallet.Domain
{
    public class Wallet : AggregateRoot
    {
        public Guid UserId { get; private set; }
        public WalletStatus Status { get; private set; } = WalletStatus.Active;

        public decimal Balance { get; private set; } = 500;

        public string Currency { get; private set; } = "$";

        // private List<WalletTransaction> _walletTransactions = new List<WalletTransaction>();
        // public IReadOnlyCollection<WalletTransaction> WalletTransactions => _walletTransactions.AsReadOnly();

        public DateTime CreatedAt { get; private set; } = DateTime.UtcNow;
        public DateTime? LastUpdate { get; private set; }



        protected Wallet() : base(Guid.Empty)
        {
        }
        private Wallet(Guid id, Guid userId) : base(id)
        {
            UserId = userId;

        }

        public static Result<Wallet> Create(Guid userId)
        {

            if (userId == Guid.Empty)
            {
                return Result<Wallet>.Failure(Error.Validation("User Id Cannot be empty"));
            }
            return Result<Wallet>.Success(new Wallet(Guid.NewGuid(), userId));
        }

        public void UpdateLastUpdate()
        {
            LastUpdate = DateTime.UtcNow;
        }
        public void Activate()
        {
            Status = WalletStatus.Active;
        }
        public void Freeze()
        {
            Status = WalletStatus.Freeze;
        }

        public void Credit(decimal amount)
        {
            Balance += amount;
        }

        public void Debit(decimal amount)
        {
            // if (this.Balance < amount)
            // {
            //     throw new DomainException($"your balance is less than {amount}");
            // }
            Balance -= amount;
        }
    }
}